### PR TITLE
change(mac): remove 'Always show OSK' option

### LINKED ISSUE #1523: bug(mac): MS Word font menu or font size menu triggers OSK
**Describe the bug**
From https://community.software.sil.org/t/keyman-11-beta-feedback/1631/12 @dyacob writes

**To Reproduce**
1. Select from MS Word font menu, or font size menu
2. Keyman OSK pops up

**Expected behavior**
OSK shouldn't appear (keyboard shortcut?)

---
**Keyman for macOS**
 - OS: macOS
 - Keyman version: 11.0.200
 - Target Application: MS Word 16.16.5, Chrome 71.0.3578.98

### LINKED ISSUE #5856: bug(mac): OSK popups randomly system-wide
**Describe the bug**
OSK popups randomly system-wide

**To Reproduce**
OSK popups randomly system-wide, it can happen anywhere. It is hard to reproduce.
But it is frequent

**Expected behavior**
OSK should not popup

**Keyman for Windows/macOS/Linux:**
 - OS: macOS Monterey
 - Keyman version: 15.0.136-alpha
 - Target Application: system-wide

### LINKED ISSUE #8965: bug(mac): Inconsistent behavior of the OSK after enabling 'Always show on-screen keyboard' option
### Describe the bug

 I noticed that after enabling 'Always show on-screen keyboard' option, the keyboard which is unchecked from the Keyboard Configuration dialog, is still enabled in the Keyman menu list and the OSK appears on the Screen.

### Reproduce the bug

1. Install 17.0.120-alpha build. 
2. Open Keyman Configuration dialog. 
3. Install any keyboard. (eg., Amharic keyboard)
4. Click the Options tab. 
5. Enable 'Always show on-screen keyboard' option. 
6. Click the Keyboards tab. 
7. Uncheck 'Amharic' Keyboard checkbox. 
8. Switch to default keyboard. 
9. Revert to Keyman Keyboard .

Here, I noticed that the Amharic keyboard is still enabled in the Keyboard menu list. And the OSK still appears on the screen. Seems to be an issue. 

   

### Expected behavior

After disabling the Keyboard in Keyboards view, the disabled keyboard should not appear in the keyboard menu list. 

### Related issues

_No response_

### Keyman apps

- [ ] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [X] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

17.0.120-alpha build

### Operating system

macOS 13.4 Ventura

### Device

_No response_

### Target application

_No response_

### Browser

_No response_

### Keyboard name

_No response_

### Keyboard version

_No response_

### Language name

_No response_

### Additional context

_No response_

### LINKED ISSUE #12341: feat(mac): always save OSK window size and location and apply when reopening
### Is your feature request related to a problem? Please describe.

It should be standard behavior to remember how the user customized their windows -- so some might consider this a bug rather than a feature.

### Describe the solution you'd like

When the on-screen keyboard window is moved or resized, Keyman should save the size and location. When the OSK is closed and opened again later, the user should see it appear where they last specified.

If the saved location or size causes the OSK to be displayed off-screen, then it should be adjusted so that it is visible.

### Describe alternatives you've considered

It's possible that some users would want the size and location of the window to vary based on the keyboard in use, but that may be confusing and require more UI design to make this association clear to the user. 

### Related issues

_No response_

### Keyman apps

- [ ] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [X] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

_No response_

### Operating system

_No response_

### Device

_No response_

### Target application

_No response_

### Browser

_No response_

### Keyboard name

_No response_

### Keyboard version

_No response_

### Language name

_No response_

### Additional context

_No response_

### LINKED ISSUE #12342: change(mac): remove option 'Always show  on-screen keyboard'
### Describe the bug

The option 'Always show on-screen keyboard' is meant to determine whether the OSK should be opened or closed when a user switches in and out of Keyman. Instead of managing this state with an option, simply remember the shown/hidden state of the OSK as set by the user and apply it whenever necessary.

Default to the OSK not being shown when Keyman is first used after installation.

### Reproduce the bug

_No response_

### Expected behavior

_No response_

### Related issues

#5856 
#1523 
#433

### Keyman apps

- [ ] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [X] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

_No response_

### Operating system

_No response_

### Device

_No response_

### Target application

_No response_

### Browser

_No response_

### Keyboard name

_No response_

### Keyboard version

_No response_

### Language name

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sgschantz**
Remove the option in the Options pane of the Configuration titled 'Always show on-screen keyboard'.

Instead, automatically track the OSK state (shown or hidden) and apply it whenever Keyman is activated. Also, save the size and location of the OSK window.

Fixes #1523
Fixes #5856 
Fixes #8965 
Fixes #12341 
Fixes #12342 

# User Testing

* GROUP_CURRENT_OS: Use macOS Sonoma
* GROUP_OLD_OS: Use macOS High Sierra (10.13), Mojave (10.14) or Catalina (10.15)

## TEST_ALWAYS_SHOW_OPTION_IS_GONE

1. Open the Configuration window and go to the Options pane
1. Verify that the checkbox labeled 'Always show on-screen keyboard' is gone

## TEST_OSK_REMEMBERS_SHOW_HIDE_STATE_DURING_SESSION

1. Open the OSK while using Keyman
1. Change from Keyman to another input method such as the U.S. Keyboard provided by Apple
1. Confirm that the OSK automatically hides
1. Switch back to the Keyman input method
1. Confirm that the OSK reappears

## TEST_OSK_STABILITY

1. Open the OSK while using Keyman
1. Change between text input applications with OSK open
1. Click on menus in the various applications
1. Confirm that the OSK does not flicker or disappear while switching apps or using menus

## TEST_OSK_REMAINS_HIDDEN_WITH_KEYMAN_INACTIVE

1. Open the OSK while using Keyman
1. Change from Keyman to another input method such as the U.S. Keyboard provided by Apple
1. Switch applications and click on various menus and windows
1. Confirm that the Keyman OSK never flashes on screen while Keyman is inactive

## TEST_OSK_REMEMBERS_SHOW_HIDE_STATE_AFTER_RESTART

1. Open the OSK while using Keyman
1. Restart the machine
1. Switch to the Keyman input method
1. Confirm that the OSK appears

## TEST_OSK_REMEMBERS_SIZE_AND_LOCATION_AFTER_RESTART

1. Open the OSK while using Keyman
1. Change the size and location of the OSK to something
1. Restart the machine
1. Switch to the Keyman input method
1. Confirm that the OSK appears at the same size and location as when last used

### CODE REVIEW

**Comment by mcdurdin** [comment]
> change(mac): remove 'Always show OSK' option

This is more than a change -- it's a significant fix to the inconsistent and weird popup of the OSK. So we should embed that into the title. e.g.

> fix(mac): make OSK display consistent and remove redundant 'Always show OSK' option

File: mac/Keyman4MacIM/Keyman4MacIM/KMConfiguration/KMConfigurationWindowController.m [hunk 1]
```diff
@@ -14,7 +14,6 @@
 @interface KMConfigurationWindowController ()
```
Comment (mcdurdin) [inline mac/Keyman4MacIM/Keyman4MacIM/KMConfiguration/KMConfigurationWindowController.m:14]: This dialog is starting to feel a little sparse!

File: mac/Keyman4MacIM/Keyman4MacIM/KMInputController.m [hunk 2]
```diff
@@ -59,102 +61,44 @@ - (void)handleBackspace:(NSEvent *)event {
...
- _kmModes = [[NSDictionary alloc] initWithObjectsAndKeys:modeList, kTSInputModeListKey,
- modeOrder, kTSVisibleInputModeOrderedArrayKey, nil];
- }
- 
- return _kmModes;
- }
- */
+- (void)deactivateServer:(id)sender {
+  [KMInputMethodLifecycle.shared deactivateClient:sender];
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
```
Comment (mcdurdin) [inline mac/Keyman4MacIM/Keyman4MacIM/KMInputController.m:93]: This doesn't seem to have a corresponding addObserver in activateServer? Perhaps a comment may clarify

File: mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodLifecycle.m [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+ * Created by Shawn Schantz on 2024-09-09.
+ * 
+ * This class is responsible for determining the state of the Keyman input method.
+ * It is called from the KMInputController (a subclass of IMKInputController), and
+ * shares changes in the state of the input method by synchronously posting
+ * notifications to NSNotificationCenter.
+ */
+
+/**
+ * This class is needed because many activateServer and deactivateServer messages sent from macOS
```
Comment (mcdurdin) [inline mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodLifecycle.m:13]: :+1: great commentary

**Review by mcdurdin — APPROVED** [review]
LGTM, this is a really clean update, look forward to seeing it in action

**Comment by dinakaranr** [comment]
# Test Results

### GROUP_CURRENT_OS: Use macOS Sonoma
I tested this issue with the attached "Keyman 18.0.110-alpha-local" build on the MacOS Catalina. Installed the Keyman keyboard package version 18.0.110.dmg

* **TEST_ALWAYS_SHOW_OPTION_IS_GONE (Passed):** 
1. Open the configuration dialog.
2. Installed the "Khmer" and "Tamil99" keyboards.(e.g. IPA SIL)
3. Click on the "Options" tab.
4. Verified that the checkbox labeled "Always show on-screen keyboard" is gone.

* **TEST_OSK_REMEMBERS_SHOW_HIDE_STATE_DURING_SESSION (Passed):** 
1. Open the OSK by clicking the keyman icon on the toolbar.
2. Open the "Text Editor" application.
3. Select the "Khmer" keyboard by clicking the keyman icon on the toolbar.
4. Type some texts, and then text reflected it.
5. Verified that the OSK appeared.
6. Change the keyboard from Keyman to Apple(U.S. Keyboard)
7. Verified that the OSK disappeared automatically.
8. Change the keyboard from Apple(U.S. Keyboard) to Keyman.
9. Verified that the OSK reappeared automatically.

* **TEST_OSK_STABILITY (Passed):** 
1. Open the OSK by clicking the keyman icon on the toolbar.
2. Change the keyboard from Keyman to Apple(U.S. Keyboard)
3. Verified that the OSK appeared.
4. Open the "Text Editor" or "Notes" application.
5. Click on the app's menu and items.
6. Verified that the OSK did not flicker.

* **TEST_OSK_REMAINS_HIDDEN_WITH_KEYMAN_INACTIVE (Passed):** 
1. Open the OSK by clicking the keyman icon on the toolbar.
2. Change the keyboard from Keyman to Apple(U.S. Keyboard).
3. Verified that the OSK appeared.
4. Switch the application from "Text Editor" to "Stickes."
5. Click on the app's menu and items.
6. Verified that the OSK does not flash when it is not inactive state.

* **TEST_OSK_REMEMBERS_SHOW_HIDE_STATE_AFTER_RESTART (Passed):** 
1. Open the OSK by clicking the keyman icon on the toolbar.
2. Restart the machine.
3. Verified that the keyman icon appears on the toolbar.
4. Switch to the keyman keyboard 
5. Verified the OSK keyboard is changed.
6. Verified that the OSK appeared.

* **TEST_OSK_REMEMBERS_SIZE_AND_LOCATION_AFTER_RESTART (Passed):**
1. Open the OSK by clicking the keyman icon on the toolbar.
2. Change the OSK size and move the OSK to another place on the monitor screen.
3. Verified that the OSK appeared in a bigger size.
4. Restart the machine.
5. Verified that the keyman icon appears on the toolbar.
6. Switch to the keyboard. 
7. Verified the OSK keyboard is changed.
8. Verified that the OSK appeared in the same bigger size and the OSK position appeared in the same place earlier too.

### GROUP_OLD_OS: Use macOS Catalina (10.15)
I tested this issue with the attached "Keyman 18.0.110-alpha-local" build on the MacOS Catalina. Installed the Keyman keyboard package version 18.0.110.dmg

* **TEST_ALWAYS_SHOW_OPTION_IS_GONE (Passed):** 
1. Open the configuration dialog.
2. Installed the "Khmer" and "Tamil99" keyboards.(e.g. IPA SIL)
3. Click on the "Options" tab.
4. Verified that the checkbox labeled "Always show on-screen keyboard" is gone.

* **TEST_OSK_REMEMBERS_SHOW_HIDE_STATE_DURING_SESSION (Passed):** 

1. Open the OSK by clicking the keyman icon on the toolbar.
2. Open the "Text Editor" application.
3. Select the "Khmer" keyboard by clicking the keyman icon on the toolbar.
4. Type some texts, and then text reflected it.
5. Verified that the OSK appeared.
6. Change the keyboard from Keyman to Apple(U.S. Keyboard)
7. Verified that the OSK disappeared automatically.
8. Change the keyboard from Apple(U.S. Keyboard) to Keyman.
9. Verified that the OSK reappeared automatically.

* **TEST_OSK_STABILITY (Passed):** 
1. Open the OSK by clicking the keyman icon on the toolbar.
2. Change the keyboard from Keyman to Apple(U.S. Keyboard)
3. Verified that the OSK appeared.
4. Open the "Text Editor" or "Notes" application.
5. Click on the app's menu and items.
6. Verified that the OSK did not flicker.

* **TEST_OSK_REMAINS_HIDDEN_WITH_KEYMAN_INACTIVE (Passed):** 
1. Open the OSK by clicking the keyman icon on the toolbar.
2. Change the keyboard from Keyman to Apple(U.S. Keyboard).
3. Verified that the OSK appeared.
4. Switch the application from "Text Editor" to "Stickes."
5. Click on the app's menu and items.
6. Verified that the OSK does not flash when it is not inactive state.

* **TEST_OSK_REMEMBERS_SHOW_HIDE_STATE_AFTER_RESTART (Passed):** 
1. Open the OSK by clicking the keyman icon on the toolbar.
2. Restart the machine.
3. Verified that the keyman icon appears on the toolbar.
4. Switch to the keyman keyboard 
5. Verified the OSK keyboard is changed.
6. Verified that the OSK appeared.

* **TEST_OSK_REMEMBERS_SIZE_AND_LOCATION_AFTER_RESTART (Passed):**
1. Open the OSK by clicking the keyman icon on the toolbar.
2. Change the OSK size and move the OSK to another place on the monitor screen.
3. Verified that the OSK appeared bigger.
4. Restart the machine.
5. Verified that the keyman icon appears on the toolbar.
6. Switch to the keyboard. 
7. Verified the OSK keyboard is changed.
8. Verified that the OSK appeared in the same bigger size and the OSK position appeared in the same place earlier too.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.114-alpha](https://keyman.com/downloads/releases/alpha/18.0.114)

**Pull request merged**